README.md
# Service-it

[![Version status](https://img.shields.io/pypi/status/serviceit?label=status)](https://pypi.org/project/serviceit)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![Python version compatibility](https://img.shields.io/pypi/pyversions/serviceit?label=Python)](https://pypi.org/project/serviceit)
[![Version on GitHub](https://img.shields.io/github/v/release/dmyersturnbull/service-it?include_prereleases&label=GitHub)](https://github.com/dmyersturnbull/service-it/releases)
[![Version on PyPi](https://img.shields.io/pypi/v/serviceit?label=PyPi)](https://pypi.org/project/serviceit)  
[![Build (Actions)](https://img.shields.io/github/workflow/status/dmyersturnbull/service-it/Build%20&%20test?label=Tests)](https://github.com/dmyersturnbull/service-it/actions)
[![Coverage (coveralls)](https://coveralls.io/repos/github/dmyersturnbull/service-it/badge.svg?branch=main&service=github)](https://coveralls.io/github/dmyersturnbull/service-it?branch=main)
[![Maintainability (Code Climate)](https://api.codeclimate.com/v1/badges/6b804351b6ba5e7694af/maintainability)](https://codeclimate.com/github/dmyersturnbull/service-it/maintainability)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/dmyersturnbull/service-it/badges/quality-score.png?b=main)](https://scrutinizer-ci.com/g/dmyersturnbull/service-it/?branch=main)
[![Created with Tyrannosaurus](https://img.shields.io/badge/Created_with-Tyrannosaurus-0000ff.svg)](https://github.com/dmyersturnbull/mandos)


Turn any Python function into a service that receives JSON payloads on some port.

Here’s a trivial example:

```python
import serviceit
def receiver(payload):
    print(payload)
server = serviceit.server(1533, receiver)
# Now it will receive JSON on 1533. For convenience:
server.client().send(dict(message="hi"))
print(server.bytes_processed)
```

#### More complex example: isolate code
You can use this to isolate a component of you code.
For example, rdkit can be installed through Conda but not Pip (or Poetry).
So, create a service and import it in an Anaconda environment to create a server,
and in your pip-installed client code.

**In a Conda environment**, create a service that listens on port 1533:

```python
import serviceit

def _receiver(payload):
    # noinspection PyUnresolvedReferences
    from rdkit.Chem.inchi import InchiToInchiKey
    inchikey = InchiToInchiKey(payload["inchi"])
    print(inchikey)

server = serviceit.server(1533, _receiver)
```

**On your pip-install client side:**

```python
import serviceit
client = serviceit.client(1533)
client.send(dict(inchi="InChI=1S/H2O/h1H2"))
```


[New issues](https://github.com/dmyersturnbull/service-it/issues) and pull requests are welcome.
Please refer to the [contributing guide](https://github.com/dmyersturnbull/service-it/blob/master/CONTRIBUTING.md).  
Generated with [Tyrannosaurus](https://github.com/dmyersturnbull/tyrannosaurus).


CHANGELOG.md
# Changelog for service-it

Adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
and [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [0.2.1] - 2021-06-08

### Fixed
- Updated build

## [0.2.0] - 2021-04-17

### Removed
- Support for Python 3.7

### Fixed
- Dropped docs
- Removed dockerfile
- Improved CI/CD
- Bumped versions in lock file


## [0.1.0] - 2020-05-26

### Added:
- Initial code.


tests/test_cli.py
import pytest

from serviceit import __copyright__
from serviceit import cli

class TestCli:

    def test_cli(self):
        response = cli.info()
        assert __copyright__ in response


if __name__ == "__main__":
    pytest.main()


tests/test_init.py
import pytest
from datetime import datetime

import serviceit


def wait(slow):
    for i in list(range(0, slow)):
        if i % slow / 10 == 0:
            i += 1


class TestInit:
    def test_passing(self):
        # TODO This is stochastic and depends on this parameter
        slow = int(1e6)
        port = 0
        data = []
        then = datetime.now()

        def process(d):
            data.append(d)
            wait(slow)

        server = serviceit.server(port, process)
        client = serviceit.client(server.port)
        assert server.last_processed is None
        assert client.last_sent is None
        assert client.payloads_sent == 0
        assert server.payloads_processed == 0
        client.send(dict(yes=True))
        wait(slow)
        assert server.payloads_processed == 1
        client.send(dict(yez=True))
        wait(slow)
        # assert server.payloads_processed() == 2
        client.send(dict(yek=True))
        assert client.payloads_sent == 3
        # give it sec to poll and process
        wait(slow)
        # while server.payloads_processed() < 2:
        #    pass
        assert str(data) == str([{"yes": True}, {"yez": True}, {"yek": True}])
        assert client.bytes_sent == 13 * 3
        # TODO this works, but only if we set slow=int(8*1E7)
        # assert server.bytes_processed == 13*3
        assert server.last_processed is not None
        assert client.last_sent is not None
        assert then < client.last_sent
        assert then < client.last_sent
        # TODO I think this isn't guaranteed anymore. True?
        # assert client.last_sent <= server.last_processed
        client.send(dict(yes=True))
        assert client.payloads_sent == 4
        wait(slow)
        # while server._service.server.payloads_processed < 2:
        #    pass

        assert str(data) == str([{"yes": True}, {"yez": True}, {"yek": True}, {"yes": True}])
        assert server.payloads_processed == 4


if __name__ == "__main__":
    pytest.main()


tests/test_server.py
import pytest

from serviceit.server import ServiceServer, Payload

from .test_init import wait


class TestService:

    def test_payload(self):
        input = {"testing": "123"}
        assert str(Payload.encode(input).decode()).replace('"', "'") == str(input).replace('"', "'")

    def test_server_client(self):
        data = []

        def add(d):
            data.append(d)

        server = ServiceServer(add, 0)
        client = server.client()
        client.send(dict(msg="hi"))
        # give it sec to poll and process
        while server.bytes_processed == 0:
            pass
        assert str(data) == str([dict(msg="hi")])

    def test_close_client(self):
        data = []

        def add(d):
            data.append(d)

        server = ServiceServer(add, 0)
        client = server.client()
        client.send(dict(msg="hi"))
        assert client.is_open
        client.close()
        assert not client.is_open
        client.send({})

    def test_close_server(self):
        data = []

        def add(d):
            data.append(d)

        server = ServiceServer(add, 0)
        client = server.client()
        client.send(dict(msg="hi"))
        assert server.is_open
        server.close()
        assert not server.is_open
        client.send({})

    def test_str(self):
        server = ServiceServer(lambda s: s, 0)
        client = server.client()
        assert str(server).startswith('ServiceServer') and '127.0.0.1' in str(server)
        assert repr(server).startswith('ServiceServer') and '127.0.0.1' in repr(server)
        assert str(client).startswith('ServiceClient') and 'localhost' in str(client)
        assert repr(client).startswith('ServiceClient') and 'localhost' in repr(client)


    def test_receive_none(self):
        slow = int(1e6)
        data = []

        def add(d):
            data.append(d)

        server = ServiceServer(add, 0)
        client = server.client()
        client.send(dict(msg="hi"))
        wait(slow)
        assert client.receive() is None

    """
    # TODO
    def test_receive_some(self):
        slow = int(1e6)
        data = []

        def add(d):
            data.append(d)
            return {'success': True}

        server = ServiceServer(add, 0)
        client = server.client()
        client.send(dict(msg="hi"))
        wait(slow)
        assert client.receive() == {'success': True}
    """


if __name__ == "__main__":
    pytest.main()


serviceit/server.py
"""
Main code for service-it.
Contains the ``ServiceClient`` and ``ServiceServer`` classes.
"""
from __future__ import annotations
import json
import abc
import threading
import logging
import socket
import socketserver
from datetime import datetime
from typing import Mapping, Callable, Any, Optional, Tuple, Type

Json = Mapping[Any, Any]
Responder = Callable[[Mapping[Any, Any]], Optional[Json]]
logger = logging.getLogger("serviceit")


class Payload(dict):
    """
    A JSON payload.
    """

    @classmethod
    def decode(cls, bts: bytes) -> Payload:
        payload = bts.decode("utf8")
        try:
            return Payload(json.loads(payload))
        except json.decoder.JSONDecodeError:
            logger.error("Failed on payload: {}".format(payload))
            raise

    def encode(self) -> bytes:
        return bytes(json.dumps(self), encoding="utf8")


class Handler(socketserver.BaseRequestHandler):
    def handle_inner(self) -> bytes:
        req = self.request.recv(4096)
        if len(req) == 0:
            logger.debug("Received an empty payload.")
            return req
        logger.info("Received payload of {} bytes. Processing...".format(len(req)))
        payload = Payload.decode(req)
        returned = self.receive(payload)
        logger.debug("Payload was {}".format(payload))
        logger.debug("Responding with {}".format(returned))
        if returned is not None:
            response = Payload(returned).encode()
            self.request.send(response)
            logger.info("Processed payload and replied.")
        else:
            logger.info("Processed payload.")
        return req

    @abc.abstractmethod
    def receive(self, payload: Json) -> Optional[Json]:
        raise NotImplementedError()


class Server(socketserver.ThreadingTCPServer):
    # noinspection PyAbstractClass
    def __init__(self, server_address, handler_class: Type[Handler]):
        self.bytes_processed = [0]

        class MyHandler(handler_class):
            # noinspection PyMethodParameters
            def handle(inner_self):
                req = inner_self.handle_inner()
                self.bytes_processed[0] += len(req)

        super().__init__(server_address, MyHandler)
        self.last_received = None
        self.last_processed = None
        self.payloads_received = 0
        self.payloads_processed = 0
        self._first = False
        self.started_at = datetime.now()

    def process_request(self, request, client_address: Tuple[str, int]) -> None:
        # this is a very weird workaround
        # but the first call seems to make two requests
        if self._first:
            self.payloads_received += 1
            self.last_received = datetime.now()
        super().process_request(request, client_address)
        if not self._first:
            self._first = True
        else:
            self.payloads_processed += 1
            self.last_processed = datetime.now()

    def handle_error(self, request, client_address):
        logger.error("Failed processing request from {}".format(client_address), exc_info=True)


class ServiceClient:
    """
    A socket to a socketserver (``ServiceServer``) that receives JSON payloads.
    See ``serviceit.client`` for more info.
    """

    def __init__(
        self, port: int, socket_family: int = socket.AF_INET, socket_type: int = socket.SOCK_STREAM
    ):
        """
        Opens a new socket.
        Refer to ``serviceit.client``.

        Args:
            port: The port to send payloads. Must be a positive integer.
            socket_family: A bit flag. See the constructor for ``socket.socket``.
            socket_type: A bit flag. See the constructor for ``socket.socket``.
        """
        self.ip = "localhost"
        self.port = port
        self._last_sent = None
        self._payloads_sent = 0
        self._bytes_sent = 0
        self._socket = socket.socket(socket_family, socket_type)
        self._connect()
        self._open = True

    @property
    def payloads_sent(self) -> int:
        """
        Get N payloads sent.

        Returns:
            The number of payloads sent. Always up-to-date.
        """
        return self._payloads_sent

    @property
    def bytes_sent(self) -> int:
        """
        Get N bytes sent.

        Returns:
            The number of bytes sent. Always up-to-date.
        """
        return self._bytes_sent

    @property
    def last_sent(self) -> Optional[datetime]:
        """
        Get datetime of last packet.

        Returns:
            The datetime when the last packet was sent, or None if none were sent. Always up-to-date.
        """
        return self._last_sent

    def send(self, data: Json) -> None:
        """
        Sends a JSON payload to the server.

        Args:
            data: An arbitrary dict (Any to Any).
                  If you have a list (``[ {} ]``), loop and send each element instead.
        """
        self._connect()
        encoded = Payload(data).encode()
        self._socket.send(encoded)
        self._last_sent = datetime.now()
        logger.debug("Sent {} bytes".format(len(encoded)))
        self._payloads_sent += 1
        self._bytes_sent += len(encoded)

    def receive(self) -> Optional[Json]:
        """
        Reads data sent back from the server.
        WARNING: This may be out-of-order.
        WARNING: Be careful with this method, which may block.

        Returns:

            The response received as a dict (Any to Any).
        """
        rec = self._socket.recv(4096)
        if len(rec) == 0:
            return None
        else:
            return Payload.decode(rec)

    @property
    def is_open(self) -> bool:
        """
        Return whether this socket is open.

        Returns:
            Whether ``self.close()`` was called.
            It's possible for the socket to be closed even if this returns ``True``.
        """
        return self._open

    def close(self) -> None:
        """
        Closes this socket.
        """
        self._socket.close()
        self._open = False

    def _connect(self):
        self._socket.close()
        self._open = False
        self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._socket.connect((self.ip, self.port))
        self._open = True

    def __repr__(self) -> str:
        """
        Returns a representation string, which has the memory address.
        """
        return "{}@{}:{}:{}(sent={} @ {})".format(
            self.__class__.__name__,
            self.ip,
            self.port,
            "open" if self._open else "closed",
            self.payloads_sent,
            hex(id(self)),
        )

    def __str__(self) -> str:
        """
        Returns a human-readable string.
        """
        return "{}@{}:{}:{}(sent={})".format(
            self.__class__.__name__,
            self.ip,
            self.port,
            "open" if self._open else "closed",
            self.payloads_sent,
        )


class ServiceServer:
    """
    A socketserver that receives JSON payloads.
    See ``serviceit.server`` for more info.
    """

    def __init__(
        self,
        receiver: Optional[Callable[[Json], Optional[Json]]],
        port: int,
        poll_interval: float = 0.001,
    ):
        """
        Creates a new ``ServiceServer``, which listens on a new thread and spawns an additional thread per request.
        Refer to ``serviceit.server``.
        """

        class H(Handler):
            def receive(self, payload: Json):
                receiver(payload)

        ip = "localhost"
        self.poll_interval = poll_interval
        self._handler_class = H
        self._server = Server((ip, port), self._handler_class)
        self.ip, self.port = self._server.server_address
        self._server_thread = threading.Thread(
            target=self._server.serve_forever, kwargs=dict(poll_interval=poll_interval)
        )
        self._server_thread.setDaemon(True)
        self._server_thread.start()
        self._open = True
        logger.info("Started service at {}:{}.".format(self.ip, self.port))

    @property
    def started_at(self) -> datetime:
        """
        Get when the server started.

        Returns:
            The datetime the server was created.
        """
        return self._server.started_at

    @property
    def last_received(self) -> Optional[datetime]:
        """
        Get when the last packet was received.

        Returns:
            The datetime of the last payload received and retrieved by polling, or None if no payloads were received.
            This value is not guaranteed to be up-to-date, since the statistic lives on another thread.
        """
        return self._server.last_received

    @property
    def last_processed(self) -> Optional[datetime]:
        """
        Get when the last package finished processing.

        Returns:
            The datetime of the last payload for which processing finished, or None if no payloads were processed.
            This value is not guaranteed to be up-to-date, since the statistic lives on another thread.
        """
        return self._server.last_processed

    @property
    def payloads_received(self) -> int:
        """
        Get the total N payloads received.

        Returns:
            The number of JSON payloads received and retrieved by polling.
            This number is not guaranteed to be up-to-date, since the statistic lives on another thread.
        """
        return self._server.payloads_received

    @property
    def payloads_processed(self) -> int:
        """
        Get the total N payloads processed.

        Returns:
            The number of JSON payloads for which processing finished.
            This number is not guaranteed to be up-to-date, since the statistic lives on another thread.
        """
        return self._server.payloads_processed

    @property
    def bytes_processed(self) -> int:
        """
        Get the total N bytes processed.

        Returns:
            The number of bytes processed.
            This number is not guaranteed to be up-to-date, since the statistic lives on another thread.
            In practice, it seems to lag behind ``payloads_processed``.
        """
        return self._server.bytes_processed[0]

    def client(self) -> ServiceClient:
        """
        Opens a socket to this server.
        In general, you would want to get a client from somewhere without access to this ``ServiceServer``.
        In some ways, this method is a little pointless, but it's useful for testing.

        Returns:
            A new ``ServiceClient`` instance.
        """
        return ServiceClient(self.port)

    @property
    def is_open(self) -> bool:
        """
        Returns whether the socket is open.

        Returns:
            In theory, whether this server is accepting connections.
            Specifically, whether ``self.close()`` was called.
            In practice, this may be out-of-date, and the server may have died in other ways.
        """
        return self._open

    def close(self) -> None:
        """
        Shuts down this server, closing the connect.
        WARNING: Currently does not stop the polling thread.
        """
        self._open = False
        self._server.shutdown()
        # TODO
        # self._server_thread.

    def __repr__(self) -> str:
        """
        Returns a representation string, which has the memory address.
        """
        return "{}@{}:{}:{}(started={},processed={} @ {})".format(
            self.__class__.__name__,
            self.ip,
            self.port,
            "open" if self._open else "closed",
            self.started_at,
            self.payloads_processed,
            hex(id(self)),
        )

    def __str__(self) -> str:
        """
        Returns a human-readable string.
        """
        return "{}@{}:{}:{}(started={},processed={})".format(
            self.__class__.__name__,
            self.ip,
            self.port,
            "open" if self._open else "closed",
            self.started_at,
            self.payloads_processed,
        )


__all__ = ["ServiceServer", "ServiceClient", "Json", "Responder"]


serviceit/__init__.py
"""
Metadata for service-it.
"""

from pathlib import Path
import logging

from importlib.metadata import PackageNotFoundError
from importlib.metadata import metadata as __load

from serviceit.server import ServiceServer, ServiceClient, Responder

pkg = Path(__file__).absolute().parent.name
logger = logging.getLogger(pkg)
_metadata = None
try:
    metadata = __load(Path(__file__).absolute().parent.name)
    __status__ = "Development"
    __copyright__ = "Copyright 2020–2021"
    __date__ = "2020-08-24"
    __uri__ = metadata["home-page"]
    __title__ = metadata["name"]
    __summary__ = metadata["summary"]
    __license__ = metadata["license"]
    __version__ = metadata["version"]
    __author__ = metadata["author"]
    __maintainer__ = metadata["maintainer"]
    __contact__ = metadata["maintainer"]
except PackageNotFoundError:
    logger.error("Could not load metadata for serviceit. Is it not installed?")


def server(port: int, receiver: Responder, poll_interval: float = 0.001) -> ServiceServer:
    """
    Starts a new threaded socketserver on localhost that accepts JSON.
    The server will run on a new thread (python ``threading``), daemonized, and will spawn a new thread per request.

    Args:
        port: If 0, lets the kernel choose the port, which will then be accessible with ``my_server.port``.
        receiver: A function that receives JSON as a dict, processes the request.
                  May return ``None`` or a dict (Any to Any), which will be sent back to the client.
                  Note that the client will not be able to tell which packet packet the server is responding to.
                  In fact, it may even be from a different client.
                  Therefore, you should always include the original payload or a hash of it.
                  A good choice might be ``return dict(success=True, payload=payload)``.
                  The server will log requests, so there's no need to do that in ``receiver``.
        poll_interval: How often the server should poll for new payloads, in seconds

    Returns:
        A ``ServiceServer`` instance.
        It needs to be kept intact while the server accepts requests. Be nice to it.
        It will record statistics, including the number of payloads received and processed,
        and the number of bytes processed.
    """
    return ServiceServer(receiver, port, poll_interval=poll_interval)


def client(port: int) -> ServiceClient:
    """
    Opens a socket to the specified port on localhost.
    To work around problems, the socket will be closed and reopened per request.

    Args:
        port: The port on localhost. It must be a positive integer.

    Returns:
        A ``ServiceClient`` instance.
        You can call ``client.send(dictionary)`` to send JSON payloads.
        It will record statistics, including the number of payloads and bytes sent.
    """
    if port == 0:
        raise ValueError("Cannot use port==0 (let kernel choose) when creating a client")
    return ServiceClient(port)


__all__ = ["ServiceServer", "ServiceClient", "Responder", "server", "client"]
